src/audio/buffer.c

```diff
@@ -56,21 +56,46 @@ struct comp_buffer *buffer_new(struct sof_ipc_buffer *desc)
 	assert(!memcpy_s(&buffer->ipc_buffer, sizeof(buffer->ipc_buffer),
 		       desc, sizeof(*desc)));
 
-	buffer->size = desc->size;
-	buffer->alloc_size = desc->size;
-	buffer->w_ptr = buffer->addr;
-	buffer->r_ptr = buffer->addr;
-	buffer->end_addr = buffer->addr + buffer->ipc_buffer.size;
-	buffer->free = buffer->ipc_buffer.size;
-	buffer->avail = 0;
-
-	buffer_zero(buffer);
+	buffer_init(buffer, desc->size);
 
 	spinlock_init(&buffer->lock);
 
 	return buffer;
 }
 
+int buffer_set_size(struct comp_buffer *buffer, uint32_t size)
+{
+	void *new_ptr = NULL;
+	struct sof_ipc_buffer *desc = NULL;
+
+	/* validate request */
+	if (size == 0 || size > HEAP_BUFFER_SIZE) {
+		trace_buffer_error("resize error: size = %u is invalid", size);
+		return -EINVAL;
+	}
+
+	desc = &buffer->ipc_buffer;
+	if (!desc) {
+		trace_buffer_error("resize error: invalid buffer desc");
+		return -EINVAL;
+	}
+
+	new_ptr = rbrealloc(buffer->addr, RZONE_BUFFER, desc->caps, size);
```

Comment:
We need additional logic, because there are three potential cases here:
1. `size == desc->size`: we shouldn't do anything.
2. `size > desc->size`: try to realloc and fail if `!new_ptr` (you're doing it already).
3. `size < desc->size`: try to realloc and if `!new_ptr` just set size to smaller as it was done before (this can happen if there are no free blocks left, however we can still use the one we already have).

Reply:
@tlauda ok. updated and tried to implement the missing cases 1 & 3